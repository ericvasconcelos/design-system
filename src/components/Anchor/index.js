import React from 'react'

import dynamic from 'next/dynamic'
import NextLink from 'next/link'
import { bool, element, elementType, node, oneOf, oneOfType, string, func } from 'prop-types'
import { buttonStyles } from 'utils/getButtonStyle/data'
import { getIconNames } from 'utils/getIconNames'

import { Wrapper } from './styles'

const Icon = dynamic(() => import('../Icon/index.js').then(module => module.Icon), {
  ssr: false
})

export const Anchor = ({
  children,
  styleColor,
  href,
  icon,
  id,
  styleType,
  size,
  borderRadius,
  passHref,
  download,
  expanded,
  className,
  target,
  rel,
  title,
  onClick
}) => (
  <NextLink href={href} passHref={passHref} className={className}>
    <Wrapper
      data-testid="Link"
      styleColor={styleColor}
      className={className}
      {...(id && { id })}
      styleType={styleType}
      size={size}
      borderRadius={borderRadius}
      download={download}
      expanded={expanded}
      {...(target && { target })}
      {...(rel && { rel })}
      {...(title && { title })}
      {...(onClick && { onClick })}
    >
      {children}
      {icon && Icon && (
        <Icon icon={icon} styleColor={buttonStyles[styleType][styleColor].fontColor} />
      )}
    </Wrapper>
  </NextLink>
)

Anchor.propTypes = {
  children: oneOfType([node, element, elementType, string]),
  styleColor: oneOf(['blue', 'white', 'dark']),
  icon: oneOf([...getIconNames(), '']),
  id: string,
  href: string,
  styleType: oneOf(['primary', 'secondary', 'link', 'text']),
  size: oneOf(['small', 'medium', 'large']),
  borderRadius: oneOf(['default', 'rounded']),
  passHref: bool,
  download: bool,
  expanded: bool,
  className: string,
  target: string,
  rel: string,
  title: string,
  onClick: func
}

Anchor.defaultProps = {
  styleColor: 'blue',
  icon: '',
  href: '',
  styleType: 'link',
  size: 'medium',
  borderRadius: 'default',
  passHref: false,
  id: '',
  download: false,
  expanded: false,
  className: '',
  children: '',
  target: '',
  rel: '',
  title: '',
  onClick: null
}
