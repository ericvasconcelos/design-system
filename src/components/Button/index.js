import React, { forwardRef } from 'react'

import dynamic from 'next/dynamic'
import { bool, func, node, oneOf, string, number } from 'prop-types'
import { buttonStyles } from 'utils/getButtonStyle/data'
import { getIconNames } from 'utils/getIconNames'

import { Wrapper } from './styles'

const Icon = dynamic(() => import('../Icon/index.js').then(module => module.Icon), {
  ssr: false
})

/**
 * O botão é o componente usado para dar ação para alguma coisa
 */

export const Button = forwardRef(
  (
    {
      borderRadius,
      styleType,
      styleColor,
      children,
      disabled,
      id,
      icon,
      iconSize,
      isLoading,
      onClick,
      onBlur,
      size,
      type,
      expanded,
      ariaLabel,
      loadingText,
      className,
      tabIndex
    },
    ref
  ) => (
    <Wrapper
      ref={ref}
      type={type}
      borderRadius={borderRadius}
      styleColor={styleColor}
      data-testid="Button"
      disabled={disabled || isLoading}
      onClick={onClick}
      styleType={styleType}
      size={size}
      isLoading={isLoading}
      expanded={expanded}
      id={id}
      tabIndex={tabIndex}
      className={className}
      {...(ariaLabel && { 'aria-label': ariaLabel })}
      {...(onBlur && { onBlur })}
    >
      {children && !isLoading ? children : false}
      {loadingText && isLoading ? loadingText : null}
      {(icon || isLoading) && !loadingText && (
        <Icon
          icon={isLoading ? 'loader' : icon}
          styleColor={buttonStyles[styleType][styleColor].fontColor}
          size={iconSize}
        />
      )}
    </Wrapper>
  )
)

Button.propTypes = {
  /**
   * Conteúdo textual do componente
   */
  children: node,
  /**
   * Cor dos textos utilizados no componente
   */
  styleColor: oneOf(['blue', 'white', 'dark']),
  /**
   * Arredondamento do componente
   */
  borderRadius: oneOf(['default', 'rounded']),
  /**
   * Ícone recebido pelo componente
   */
  icon: oneOf([...getIconNames(), '']),
  /**
   * Tamanho do ícone
   */
  iconSize: oneOf([12, 16, 24, 32, 48, 64]),
  /**
   * Estado de 'carregando' do componente
   */
  isLoading: bool,
  /**
   * Estado de habilitado, caso desabilitado ele perde a função de click e
   * enter quando estiver focado
   */
  disabled: bool,
  /**
   * Função que será executada quando o componente for clicado
   */
  onClick: func.isRequired,
  /**
   * Função que será executada quando o componente perder o foco
   */
  onBlur: func,
  /**
   * Tipo do componente
   */
  styleType: oneOf(['primary', 'secondary', 'link', 'text']),
  /**
   * Tamanho do componente
   */
  size: oneOf(['small', 'medium', 'large']),
  /**
   * ID para o componente
   */
  id: string,
  /**
   * Define se o botão vai utilizar todo o espaço do pai ou se adaptará com o
   * conteúdo utilizado como children
   */
  expanded: bool,
  /**
   * Tipo de componente
   */
  type: oneOf(['button', 'submit']),
  /**
   * Texto recebido quando o componente estiver em estado de loading e NÃO receber nenhum icon
   */
  loadingText: string,
  /**
   * Opção para o componente ser utilizado dentro de um styled component
   */
  className: string,
  /**
   * Aria label do componente
   */
  ariaLabel: string,
  /**
   * TabIndex do componente
   */
  tabIndex: number
}

Button.defaultProps = {
  children: '',
  borderRadius: 'default',
  styleColor: 'blue',
  disabled: false,
  isLoading: false,
  icon: 'arrowRight',
  iconSize: 24,
  size: 'small',
  styleType: 'primary',
  id: '',
  expanded: true,
  type: 'button',
  loadingText: '',
  className: '',
  ariaLabel: '',
  onBlur: null,
  tabIndex: 0
}
