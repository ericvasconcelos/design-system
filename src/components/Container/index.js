import React from 'react'

import { node, string } from 'prop-types'

import { Wrapper } from './styles'

export const Container = ({ children, className }) => (
  <Wrapper data-testid="Container" className={className}>
    {children}
  </Wrapper>
)

Container.propTypes = {
  children: node.isRequired,
  className: string
}

Container.defaultProps = {
  className: ''
}
