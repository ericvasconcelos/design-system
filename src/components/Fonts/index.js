import styled from 'styled-components'

export const Fonts = styled.div`
  @font-face {
    font-display: swap;
    font-family: 'DM Sans';
    font-weight: 700;
    src: url('${process.env.NEXT_PUBLIC_FONT_PREFIX}/DMSans/DMSans-Bold.ttf');
  }

  @font-face {
    font-display: swap;
    font-family: 'DM Sans';
    font-weight: 500;
    src: url('${process.env.NEXT_PUBLIC_FONT_PREFIX}/DMSans/DMSans-Medium.ttf');
  }
  @font-face {
    font-display: swap;
    font-family: 'DM Sans';
    font-weight: 400;
    src: url('${process.env.NEXT_PUBLIC_FONT_PREFIX}/DMSans-Regular.ttf');
  }
`
