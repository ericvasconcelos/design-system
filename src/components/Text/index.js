import React from 'react'

import { node, oneOf, oneOfType, string } from 'prop-types'
import { getTypographyNames } from 'utils/getTypographyNames'
import { styleColorValidator } from 'utils/styleColorValidator'

import { SelectedComponent } from './styles'

export const Text = ({ tag, styleColor, styleType, children, className }) => (
  <SelectedComponent
    data-testid="Text"
    as={tag}
    className={className}
    styleColor={styleColor}
    styleType={styleType}
  >
    {children}
  </SelectedComponent>
)


Text.propTypes = {
  children: node.isRequired,
  className: string,
  tag: oneOfType([string, node]),
  styleType: oneOf(getTypographyNames()),
  styleColor: styleColorValidator
}

Text.defaultProps = {
  tag: 'p',
  styleType: 'body1',
  styleColor: 'greyCash400',
  className: ''
}
