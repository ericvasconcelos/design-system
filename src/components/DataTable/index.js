import React, { useRef, useState } from 'react'

import { Icon } from 'components/Icon'
import { string, arrayOf, shape, bool, func, number, object } from 'prop-types'

import { RowTable } from './RowTable'
import {
  WrapperTable,
  Table,
  Header,
  HeaderContent,
  IconSort,
  Pagination,
  ShowMoreButton,
  Loader,
  Text
} from './styles'

/** Componente de tabela é utilizado para exibir dados em forma de tabela */

export const DataTable = ({
  headers,
  collapse,
  data,
  numberPages,
  currentPage,
  onChangePage,
  onSort,
  showMoreItemsText,
  isLoading,
  isPageLoading,
  messageWhenEmpty,
  className
}) => {
  const tbodyRef = useRef(null)
  const theadRef = useRef(null)
  const cellRefs = useRef([])
  const [currentRowCell, setCurrentRowCell] = useState([0, 0])

  const handleFocus = (currentRow, currentCell) => {
    setCurrentRowCell([currentRow, currentCell])
  }

  const tableKeyDownHandler = (event, indexRow, indexCel) => {
    const rowsLength = tbodyRef.current.children.length
    const cellLength = theadRef.current.children.length

    const handleArrowRight = () => {
      event.preventDefault()
      if (indexCel < cellLength - 1) cellRefs.current[indexRow][indexCel + 1].focus()
    }
    const handleArrowLeft = () => {
      event.preventDefault()
      if (indexCel > 0) cellRefs.current[indexRow][indexCel - 1].focus()
    }
    const handleArrowDown = () => {
      event.preventDefault()
      if (indexRow < rowsLength - 1) cellRefs.current[indexRow + 1][indexCel].focus()
    }
    const handleArrowUp = () => {
      event.preventDefault()
      if (indexRow > 0) cellRefs.current[indexRow - 1][indexCel].focus()
    }
    const handleHome = () => {
      event.preventDefault()
      if (!event.ctrlKey) {
        return cellRefs.current[indexRow][0].focus()
      }
      cellRefs.current[0][0].focus()
    }

    const handleEnd = () => {
      event.preventDefault()
      if (!event.ctrlKey) {
        return cellRefs.current[indexRow][cellLength - 1].focus()
      }
      cellRefs.current[rowsLength - 1][cellLength - 1].focus()
    }

    const keyToAction = {
      ArrowRight: handleArrowRight,
      ArrowLeft: handleArrowLeft,
      ArrowDown: handleArrowDown,
      ArrowUp: handleArrowUp,
      Home: handleHome,
      End: handleEnd,
      Default: () => {}
    }

    return (keyToAction[event.key] || keyToAction.Default)()
  }

  return (
    <>
      <WrapperTable data-testid="DataTable" className={className}>
        <Table role="grid" aria-rowcount={data.length}>
          <thead>
            <tr ref={theadRef} role="row">
              {headers.map(header => (
                <Header
                  role="columnheader"
                  onClick={header.sortable ? () => onSort(header.key) : null}
                  sortable={header.sortable}
                  key={header.key}
                  id={`header-${header.key}`}
                >
                  <HeaderContent>
                    {header.label}
                    {header.sortable && (
                      <IconSort icon="chevronDown" styleColor="darkBlueCash500" />
                    )}
                  </HeaderContent>
                </Header>
              ))}
              {collapse.length !== 0 && <Header key="collapse" isCollapse />}
            </tr>
          </thead>
          {!isLoading && (
            <tbody ref={tbodyRef}>
              {data.map((row, index) => (
                <RowTable
                  key={row.key}
                  index={index}
                  data={row}
                  headers={headers}
                  collapse={collapse}
                  currentRowCell={currentRowCell}
                  handleFocus={handleFocus}
                  cellRefs={cellRefs}
                  tableKeyDownHandler={tableKeyDownHandler}
                />
              ))}
            </tbody>
          )}
        </Table>
        {data.length === 0 && !isLoading && messageWhenEmpty && <Text>{messageWhenEmpty}</Text>}
        {isLoading && <Loader />}
      </WrapperTable>
      {numberPages > 1 && currentPage < numberPages && (
        <Pagination>
          <ShowMoreButton disabled={isPageLoading} onClick={() => onChangePage(currentPage + 1)}>
            {isPageLoading ? (
              <Icon icon="loader" styleColor="lightBlueCash300" />
            ) : (
              `${showMoreItemsText}`
            )}
          </ShowMoreButton>
        </Pagination>
      )}
    </>
  )
}

DataTable.propTypes = {
  /** conteúdo da cabeçalho */
  headers: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired,
      sortable: bool.isRequired
    })
  ).isRequired,
  /** define se é expansivel */
  collapse: arrayOf(
    shape({
      key: string.isRequired,
      label: string.isRequired
    })
  ),
  /** dados a serem mostrados */
  // eslint-disable-next-line react/forbid-prop-types
  data: arrayOf(object).isRequired,
  /** total de linhas */
  totalRows: number.isRequired,
  /** numero de páginas (quando há paginação) */
  numberPages: number.isRequired,
  /** página atual (quando há paginação) */
  currentPage: number,
  /** quantos items mostrado por página */
  perPage: number,
  /** função chamada ao mudar de página */
  onChangePage: func,
  /** função chamada ao ordenar */
  onSort: func,
  className: string,
  /** texto mostrado na opção de mostrar mais itens */
  showMoreItemsText: string,
  /** define se está carregando */
  isLoading: bool,
  /** define se a página está carregando */
  isPageLoading: bool,
  /** mensagem quando a tabela estiver vazia */
  messageWhenEmpty: string
}

DataTable.defaultProps = {
  className: '',
  collapse: [],
  currentPage: 1,
  perPage: 10,
  onChangePage: () => {},
  onSort: () => {},
  showMoreItemsText: 'Mostrar mais',
  isLoading: false,
  isPageLoading: false,
  messageWhenEmpty: null
}
