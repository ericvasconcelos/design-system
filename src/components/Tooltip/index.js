
import React, { useState } from 'react'

import { Popper } from 'components/Popper'
import { func, node, string } from 'prop-types'

import { TooltipWrapper, ReferenceWrapper } from './styles'

/**
 * Tooltip é o componente mostrado quando o usuário faz o hover em algum elemento
 * que mostra uma dica ou instrução
 */

export const Tooltip = ({ className, placement, onMouseEnter, onMouseLeave, children }) => {
  const [show, setShow] = useState(false)
  const onMouseEnterInternal = () => {
    if (onMouseEnter) {
      onMouseEnter()
    }

    setShow(true)
  }
  const onMouseLeaveInternal = () => {
    if (onMouseLeave) {
      onMouseLeave()
    }

    setShow(false)
  }

  return (
    <Popper show={show} showArrow arrowColor="darkBlueCash400" placement={placement}>
      {reference => (
        <ReferenceWrapper
          data-testid="TooltipReference"
          onMouseEnter={onMouseEnterInternal}
          onMouseLeave={onMouseLeaveInternal}
          ref={reference}
        >
          {children[0]}
        </ReferenceWrapper>
      )}
      <TooltipWrapper className={className}>
        {children[1]}
      </TooltipWrapper>
    </Popper>
  )
}

Tooltip.propTypes = {
  /**
   * São 2 childrens
   * 1: Elemento de referência que o tooltip vai se basear
   * 2: Elemento que o tooltip vai receber como children
   */
  children: node.isRequired,
  /**
   * Posicionamento do tooltip
   */
  placement: string.isRequired,
  /**
   * Função externa chamada quando mouseEnter
   */
  onMouseEnter: func,
  /**
   * Função externa chamada quando mouseLeave
   */
  onMouseLeave: func,
  className: string
}

Tooltip.defaultProps = {
  onMouseEnter: null,
  onMouseLeave: null,
  className: ''
}
