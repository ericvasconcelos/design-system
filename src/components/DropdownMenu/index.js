import React, { useState } from 'react'

import { Button } from 'components/Button'
import { Popper } from 'components/Popper'
import PropTypes, { arrayOf, bool, string, func } from 'prop-types'

import { Wrapper } from './styles'
/**
 * DropDownMenu é um componente de
 * menu dropdown controlado por um botão.
 * O componente renderiza uma lista de elementos
 * passados por quen estancia o componente.
 */
export const DropdownMenu = ({
  id,
  buttonLabel,
  buttonIcon,
  buttonStyle,
  className,
  expanded,
  menuItems
}) => {
  const [show, setShow] = useState(false)

  const toggleVisibility = () => {
    setShow(!show)
  }

  return (
    <Popper showArrow show={show} arrowBorderColor="greyCash300">
      {reference => (
        <Button
          ref={referenceNode => {
            reference(referenceNode)
          }}
          expanded={expanded}
          icon={buttonIcon}
          styleType={buttonStyle}
          onClick={toggleVisibility}
          aria-haspopup
          className={className}
          aria-expanded={show}
          aria-describedby="initInstr"
          aria-owns={`result${id}`}
          aria-autocomplete="both"
          aria-activedescendant=""
          id={id}
          tabIndex={-1}
        >
          {buttonLabel}
        </Button>
      )}
      <Wrapper id={`result${id}`} aria-labelledby={id} role="listbox">
        {menuItems && menuItems.map(item => <li key={item.key}>{item.render()}</li>)}
      </Wrapper>
    </Popper>
  )
}

DropdownMenu.propTypes = {
  /**
   * buttonLabel é o texto que aparece
   * no botão que controla o menu
   */
  buttonLabel: string,
  /**
   * buttonIcon é o ícone que aparece
   * no botão que controla o menu
   */
  buttonIcon: string,
  /**
   * buttonStyle é o styleType
   * do botão que controla o menu
   */
  buttonStyle: string,
  /**
   * className se refere a prop className
   * do botão que controla o menu
   */
  className: string,
  /**
   * expanded se refere a prop expanded
   * do botão que controla o menu
   */
  expanded: bool,
  /**
   * id será passado ao
   * do botão que controla o menu
   */
  id: string.isRequired,
  /**
   * é uma lista de nodes
   * que farão parte do menu
   */
  menuItems: arrayOf(
    PropTypes.shape({
      key: string,
      render: func
    })
  )
}

DropdownMenu.defaultProps = {
  buttonLabel: '',
  buttonIcon: '',
  buttonStyle: 'primary',
  className: '',
  expanded: false,
  menuItems: []
}
