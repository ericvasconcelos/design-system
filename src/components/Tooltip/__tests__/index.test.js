import React from 'react'

import userEvent from '@testing-library/user-event'
import { Icon } from 'components/Icon/index.js'
import { Text } from 'components/Text'
import { styledRender } from 'utils/styledRender'

import { Tooltip } from '../index.js'

describe('Should render Tooltip component when', () => {
  it('has default props', () => {
    const { getByTestId } = styledRender(
      <Tooltip placement="bottom-start">
        <Icon icon="user">
          Some text here
        </Icon>
        <Text styleType="body2">Lorem ipsum dolor sit, amet co</Text>
      </Tooltip>
    )

    userEvent.hover(getByTestId('TooltipReference'))
    userEvent.unhover(getByTestId('TooltipReference'))

    expect(getByTestId('TooltipReference')).toMatchSnapshot()
  })

  it('receiving external functions', () => {
    const { getByTestId } = styledRender(
      <Tooltip
        placement="bottom-start"
        onMouseEnter={() => jest.fn()}
        onMouseLeave={() => jest.fn()}
      >
        <Icon icon="user">
          Some text here
        </Icon>
        <Text styleType="body2">Lorem ipsum dolor sit, amet co</Text>
      </Tooltip>
    )

    userEvent.hover(getByTestId('TooltipReference'))
    userEvent.unhover(getByTestId('TooltipReference'))

    expect(getByTestId('TooltipReference')).toMatchSnapshot()
  })
})

