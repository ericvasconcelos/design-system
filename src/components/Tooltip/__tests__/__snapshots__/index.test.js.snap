// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Tooltip component when has default props 1`] = `
.c1 {
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/user.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/user.svg);
  background-color: #A6B5BB;
  background-image: none;
  background-size: 24px;
  width: 24px;
  height: 24px;
  -webkit-mask-position-x: 0;
  mask-position-x: 0;
  -webkit-mask-size: 24px;
  mask-size: 24px;
  -webkit-mask-repeat: no-repeat;
  mask-repeat: no-repeat;
}

.c0 {
  display: inline-block;
}

<div
  class="c0"
  data-testid="TooltipReference"
>
  <div
    aria-hidden="true"
    aria-label=""
    class="c1"
    data-testid="Icon"
  />
</div>
`;

exports[`Should render Tooltip component when receiving external functions 1`] = `
.c1 {
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/user.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/user.svg);
  background-color: #A6B5BB;
  background-image: none;
  background-size: 24px;
  width: 24px;
  height: 24px;
  -webkit-mask-position-x: 0;
  mask-position-x: 0;
  -webkit-mask-size: 24px;
  mask-size: 24px;
  -webkit-mask-repeat: no-repeat;
  mask-repeat: no-repeat;
}

.c0 {
  display: inline-block;
}

<div
  class="c0"
  data-testid="TooltipReference"
>
  <div
    aria-hidden="true"
    aria-label=""
    class="c1"
    data-testid="Icon"
  />
</div>
`;
